lib/promauth/config.go

```diff
@@ -662,36 +663,78 @@ func (actx *authContext) initFromAuthorization(baseDir string, az *Authorization
 }
 
 func (actx *authContext) initFromBasicAuthConfig(baseDir string, ba *BasicAuthConfig) error {
-	if ba.Username == "" {
-		return fmt.Errorf("missing `username` in `basic_auth` section")
+	if ba.Username == "" && ba.UsernameFile == "" {
+		return fmt.Errorf("missing `username` and `username_file` in `basic_auth` section. either one should be configured")
 	}
-	if ba.PasswordFile == "" {
-		// See https://en.wikipedia.org/wiki/Basic_access_authentication
-		token := ba.Username + ":" + ba.Password.String()
-		token64 := base64.StdEncoding.EncodeToString([]byte(token))
-		ah := "Basic " + token64
-		actx.getAuthHeader = func() (string, error) {
-			return ah, nil
-		}
-		actx.authHeaderDigest = fmt.Sprintf("basic(username=%q, password=%q)", ba.Username, ba.Password)
-		return nil
+	if ba.Username != "" && ba.UsernameFile != "" {
+		return fmt.Errorf("both `username`=%q and `username_file`=%q are set in `basic_auth` section", ba.Username, ba.UsernameFile)
 	}
-	if ba.Password != nil {
+	if ba.Password != nil && ba.PasswordFile != "" {
 		return fmt.Errorf("both `password`=%q and `password_file`=%q are set in `basic_auth` section", ba.Password, ba.PasswordFile)
 	}
-	filePath := fscore.GetFilepath(baseDir, ba.PasswordFile)
-	actx.getAuthHeader = func() (string, error) {
-		password, err := fscore.ReadPasswordFromFileOrHTTP(filePath)
+	actx.getAuthHeader, actx.authHeaderDigest = fetchBasicAuthHeaderAndDigest(baseDir, ba)
+	return nil
+}
+
+func fetchBasicAuthHeaderAndDigest(baseDir string, ba *BasicAuthConfig) (func() (string, error), string) {
+	if ba.UsernameFile == "" {
```

Comment:
Currently, this function has a lot of branching due to different combinations of plain and file-based approaches to provide credentials. I'd suggest to simplify this by using the following:
```go


func fetchBasicAuthHeaderAndDigest(baseDir string, ba *BasicAuthConfig) (func() (string, error), string) {
       // Store closures in getUsername and getPassword vars in order to avoid branching
	var getUsername, getPassword func() (string, error) 
	var usernameDigest, passwordDigest string

	if ba.Username != "" {
		usernameDigest = fmt.Sprintf("username=%q", ba.Username)
		getUsername = func() (string, error) {
			return ba.Username, nil
		}
	}
	if ba.UsernameFile != "" {
		usernameDigest = fmt.Sprintf("usernameFile=%q", ba.UsernameFile)
		getUsername = func() (string, error) {
			username, err := fscore.ReadPasswordFromFileOrHTTP(fscore.GetFilepath(baseDir, ba.UsernameFile))
			if err != nil {
				return "", fmt.Errorf("cannot read username from `username_file`=%q set in `basic_auth` section: %w", ba.UsernameFile, err)
			}
			return username, nil
		}
	}
// Logic for password


	authHeader := func() (string, error) {
		username, err := getUsername()
		if err != nil {
			return "", fmt.Errorf("cannot read username from `username_file`=%q set in `basic_auth` section: %w", ba.UsernameFile, err)
		}
		password, err := getPassword()
		if err != nil {
			return "", fmt.Errorf("cannot read password from `password_file`=%q set in `basic_auth` section: %w", ba.PasswordFile, err)
		}
		token64 := generateBasicAuthEncodedToken(username, password)
		return "Basic " + token64, nil
	}
	authHeaderDigest := fmt.Sprintf("basic(%q, %q)", usernameDigest, passwordDigest)
	return authHeader, authHeaderDigest
}
```

This should make function much more linear.

Reply:
Sure. Thanks for your review! Will push the updated change